File: 0232-implement-queue-using-stacks/0232-implement-queue-using-stacks.cpp
class MyQueue {
public:
    /*
     A simulation:
     push(1) -> 1
     push(2) -> 1 2
     push(3) -> 1 2 3
     pop()   -> 1 2 3 transform to 3 2 1; then we'll get 1 on the top
    */
    
    stack<int> myQueue;

    MyQueue() {

    }
    
    void push(int x) {
        stack<int> tempQueue;
        while (!myQueue.empty()) {
            tempQueue.push(myQueue.top());
            myQueue.pop();
        }

        myQueue.push(x);
        while (!tempQueue.empty()) {
            myQueue.push(tempQueue.top());
            tempQueue.pop();
        }
    }
    
    int pop() {
        int x = myQueue.top();
        myQueue.pop();
        return x;
    }
    
    int peek() {
        return myQueue.top();
    }
    
    bool empty() {
        return myQueue.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */

File: 0232-implement-queue-using-stacks/0232-implement-queue-using-stacks.go
type MyQueue struct {
    items []int
}


func Constructor() MyQueue {
    return MyQueue{items: []int{}}
}


func (this *MyQueue) Push(x int)  {
    this.items = append(this.items, x)
}


func (this *MyQueue) Pop() int {
    val := this.items[0]
    this.items[0] = 0
    this.items = this.items[1:]
    return val
}


func (this *MyQueue) Peek() int {
    return this.items[0]
}


func (this *MyQueue) Empty() bool {
    return len(this.items) == 0
}


/**
 * Your MyQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * param_2 := obj.Pop();
 * param_3 := obj.Peek();
 * param_4 := obj.Empty();
 */